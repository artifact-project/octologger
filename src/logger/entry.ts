export const EntryTypes = {
	entry: 0,
	scope: 1,
	group: 2,
	timeMark: 3,
	timeMeasure: 4,
};

export type Entry = {
	type: number;
	badge: string;
	level: number;
	label: string;
	message: string;
	detail: any;
	meta: {
		fn: string;
		file: string;
		line: number;
		column: number;
	};
	parent: Entry;
	entries: Entry[];
}