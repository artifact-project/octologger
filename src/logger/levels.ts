export const LogLevels = {
	error: 0,
	warn: 1,
	info: 2,
	verbose: 3,
	debug: 4,
	log: 5,
	success: 6,
};

export const LogLevelsInvert = {} as {
	[level:number]: string;
};

for (let key in LogLevels) {
	LogLevelsInvert[LogLevels[key]] = key;
}

export type LogLevelTypes = keyof (typeof LogLevels);