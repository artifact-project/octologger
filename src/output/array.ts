import { OutputOptions, createOutput } from './output';
import { plainTextFormat } from '../format/plain-text';

export type ArrayOutputOptions = {
	out: string[];
	format?: OutputOptions['format'];
}

const firstIndent = '  | ';
const nextIndent = '   ';

export const arrayOutput = (options: ArrayOutputOptions) => {
	let indent = '';
	const out = options.out;
	const fakeConsole = {
		log: (...args: string[]) => {
			out.push(`${indent}${args.join(' ')}`);
		},

		group: (...args: string[]) => {
			out.push(`${indent}${args.join(' ')}`);
			indent = indent ? `${nextIndent}${indent}` : firstIndent;
		},

		groupEnd: () => {
			indent = indent === firstIndent ? '' : firstIndent.substr(nextIndent.length);
		},

		groupCollapsed: () => {
			fakeConsole.groupEnd();
		},
	};

	return createOutput({
		out: fakeConsole,
		format: options.format || plainTextFormat,
	});
}
