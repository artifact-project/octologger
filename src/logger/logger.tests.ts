import {logger} from './logger';
import { Entry, EntryMeta } from './logger.types';

beforeEach(() => {
	logger.clear();
	logger.setup({meta: false, output: null});
});

describe('core', () => {
	['add', 'log', 'info', 'done', 'warn', 'verbose', 'error'].forEach((level) => {
		describe(level, () => {
			it('support', () => {
				if (!logger.hasOwnProperty(level)) {
					throw new Error(`Level "${level}" not supproted`);
				}
			});

			it('detail', () => {
				// Add Log Record
				logger[level](level);
				expect(logger.last().detail).toEqual([level]);
			});
		});
	});
});

describe('meta', () => {
	const log: Entry[] = [];
	const toMeta = (arg: any): EntryMeta => ({file: arg[0], line: arg[1], col: arg[2]});
	
	beforeEach(() => {
		log.length = 0;
		logger.setup({output: (entry) => log.push(entry!)});
	});
	
	it('log', () => {
		const meta = ['log.ts', 123, 456];

		logger.add(meta, 'log-with-meta');

		expect(log.length).toBe(1);
		expect(log[0].detail).toEqual(['log-with-meta']);
		expect(log[0].meta).toEqual(toMeta(meta));
	});
	
	it('scope', () => {
		const meta = ['scope.ts', 123, 456];

		(logger.scope as any)(meta, 'scope-with-meta', {foo: 123});

		expect(log.length).toBe(1);
		expect(log[0].message).toEqual('scope-with-meta');
		expect(log[0].detail).toEqual({foo: 123});
		expect(log[0].meta).toEqual(toMeta(meta));
	});
});
