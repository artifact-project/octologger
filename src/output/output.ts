import { Format, nodeFromat, browserFormat } from '../format/format';
import { Entry, SCOPE } from '../logger/logger.types';

export type Output = (entry: Entry | null) => void;

export type OutputOptions = {
	out: Pick<Console, 'log' | 'group' | 'groupEnd'>;
	format: Format;
}

type OutputEntry = Entry & {
	printed?: boolean;
}

export const createOutput = (options: OutputOptions): Output => {
	const {
		out,
		format,
	} = options;

	const {log} = out;
	const debounced = {};
	let openScopes: OutputEntry[] = [];

	function print(entry: OutputEntry | null, skipPrinted?: boolean) {
		if (entry === null) {
			openScopes.forEach(() => {
				out.groupEnd();
			});
			openScopes.length = 0;
			return;
		}

		const parent = entry.parent as OutputEntry | null;
		const opened = openScopes.length;

		if (skipPrinted && entry.printed) {
			return;
		}

		if (opened > 0) {
			let idx = opened;

			while (idx--) {
				if (openScopes[idx] === parent) {
					break;
				} else {
					openScopes[idx].printed = true;
					out.groupEnd();
				}
			}

			openScopes.length = idx + 1;
		}

		if (entry.type === SCOPE) {
			if (entry.detail && entry.detail.state === 'idle') {
				entry.printed = true;
				log.apply(out, format(entry));
			} else {
				openScopes.push(entry);
				out.group.apply(out, format(entry));
			}
		} else {
			if (!skipPrinted && parent && parent.printed && parent.label !== '#root') {
				debounced[parent.cid] || (debounced[parent.cid] = setTimeout(() => {
					openScopes.forEach(() => {
						out.groupEnd();
					});
					openScopes.length = 0;

					const chain = [] as OutputEntry[];
					let cursor: OutputEntry | null = parent;

					do {
						chain.unshift(cursor!);
						cursor = cursor!.parent;
					} while (cursor && cursor.label !== '#root');

					chain.forEach(entry => {
						out.group.apply(out, format(entry));
					});

					parent.entries.forEach(entry => {
						print(entry, true);
					})

					chain.forEach(() => {
						out.groupEnd()
					});
				}));

				return;
			}

			entry.printed = true;
			log.apply(out, format(entry));
		}
	};

	return print;
};

export const nodeOutput = (options: Partial<OutputOptions> = {}) => createOutput({
	out: options.out || console,
	format: options.format || nodeFromat,
});

export const browserOutput = (options: Partial<OutputOptions> = {}) => createOutput({
	out: options.out || console,
	format: options.format || browserFormat,
});

export const universalOutput = typeof self === 'undefined' ? nodeOutput : browserOutput;
