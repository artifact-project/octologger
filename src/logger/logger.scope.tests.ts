import logger, { createScopeEntry } from '../logger/logger';
import { parseError, XError } from '../error/error';
import { LogLevels } from './levels';


beforeEach(() => {
	logger.setup({output: [], meta: true});
	logger.clear();
});

it('scope: base', () => {
	const scope = logger.scope('foo');
	expect(scope.scope().message).toBe('foo');
});

it('scope: createScopeEntry', () => {
	const scope = logger.scope(createScopeEntry(
		LogLevels.info,
		'X',
		'label',
		'msg',
		{foo: 'bar'},
	));

	expect(scope.scope().level).toBe(LogLevels.info);
	expect(scope.scope().badge).toBe('X');
	expect(scope.scope().label).toBe('label');
	expect(scope.scope().message).toBe('msg');
	expect(scope.scope().detail.info).toEqual({foo: 'bar'});
});

it('scope', function scopeTest() {
	let scopeMeta: XError;
	let startMeta = parseError(new Error);
	const scope = logger.scope('delay', {timeout: 123}, function inScope(scope) {
		scopeMeta = parseError(new Error);
		scope.log('wow');
		logger.log('Yep!');
	});
	logger.log('out of scope');

	// Root Entries
	expect(logger.getEntries().length).toEqual(2);
	expect(logger.getLastEntry().detail).toEqual(['out of scope']);

	// Scope API
	expect(typeof scope.log).toEqual('function');

	// Meta
	expect(scope.scope().meta.file).toEqual(startMeta.file);
	expect(scope.scope().meta.line).toEqual(startMeta.line + 1);

	// Info
	expect(scope.scope().detail.info).toEqual({timeout: 123});

	// Entries
	expect(scope.scope().entries.length).toEqual(2);
	expect(scope.scope().entries[0].detail).toEqual(['wow']);
	expect(scope.scope().entries[1].detail).toEqual(['Yep!']);

	// [0] In Scope Meta
	expect(scope.scope().entries[0].meta.file).toEqual(scopeMeta.file);
	expect(scope.scope().entries[0].meta.line).toEqual(scopeMeta.line + 1);
	expect(scope.scope().entries[0].meta.fn).toEqual('inScope');

	// [1] In Scope Meta
	expect(scope.scope().entries[1].meta.file).toEqual(scopeMeta.file);
	expect(scope.scope().entries[1].meta.line).toEqual(scopeMeta.line + 2);
	expect(scope.scope().entries[1].meta.fn).toEqual('inScope');
});

it('nested scopes', function inScope0() {
	const d1 = logger.scope('first', function inScope1(d1) {
		d1.log('depth-1');

		logger.scope('second', function inScope2(d2) {
			d1.log('depth-1-1');
			logger.log('depth-2');

			d2.scope('third', function inScope3(d3) {
				d3.log('depth-3');
			});
		});
	});

	expect(logger.getEntries().length).toEqual(1);

	// D1
	const s1 = d1.scope();
	expect(s1.meta.fn.split('.').pop()).toEqual('inScope0');
	expect(`s1.length:${s1.entries.length}`).toEqual('s1.length:3');

	expect(s1.entries[0].detail).toEqual(['depth-1']);
	expect(s1.entries[0].meta.fn).toEqual('inScope1');

	expect(s1.entries[1].message).toEqual('second');
	expect(s1.entries[1].meta.fn).toEqual('inScope1');

	expect(s1.entries[2].detail).toEqual(['depth-1-1']);
	expect(s1.entries[2].meta.fn).toEqual('inScope2');

	// D2
	const s2 = s1.entries[1];
	expect(s2.meta.fn).toEqual('inScope1');
	expect(`s2.length:${s2.entries.length}`).toEqual('s2.length:2');

	expect(s2.entries[0].detail).toEqual(['depth-2']);
	expect(s2.entries[0].meta.fn).toEqual('inScope2');

	expect(s2.entries[1].message).toEqual('third');
	expect(s2.entries[1].meta.fn).toEqual('inScope2');

	// D3
	const s3 = s2.entries[1];
	expect(s3.meta.fn).toEqual('inScope2');
	expect(`s3.length:${s3.entries.length}`).toEqual('s3.length:1');
	expect(s3.entries[0].detail).toEqual(['depth-3']);
	expect(s3.entries[0].meta.fn).toEqual('inScope3');
});