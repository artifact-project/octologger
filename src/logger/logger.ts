import { Entry, ScopeEntry, LoggerOptions, LoggerAPI, LoggerFactoryAPI, Logger, LoggerContext, LoggerScope, LoggerScopeContext, ContextSnapshot, EntryMeta, ScopeExecutor, SCOPE, ENTRY } from './logger.types';
import { universalOutput } from '../output/output';
import { now } from '../utils/utils';

const time = new Date();
let cid = 0;

type CompactMetaArg = [string, number, number];

function isMeta(arg: any): arg is CompactMetaArg {
	return arg && arg.length === 3 && arg[0] && arg[1] > 0 && arg[2] > 0;
}

function toMeta(arg: CompactMetaArg): EntryMeta {
	return {file: arg[0], line: arg[1], col: arg[2]};
}

export function isLogEntry(x: any): x is Entry {
	return x && x.type !== void 0 && x.level !== void 0;
}

export function createEntry(
	level: Entry['level'],
	badge: Entry['badge'],
	label: Entry['label'],
	message: Entry['message'],
	detail: Entry['detail'],
): Entry {
	return {
		cid: ++cid,
		ts: 0,
		type: ENTRY,
		level,
		badge,
		label,
		message,
		detail,
		meta: null,
		parent: null,
		entries: [],
	};
}

export function createScopeEntry(
	level: Entry['level'],
	badge: Entry['badge'],
	label: Entry['label'],
	message: Entry['message'],
	detail: Entry['detail'],
): ScopeEntry {
	return {
		cid: ++cid,
		ts: 0,
		level: level,
		type: SCOPE,
		badge,
		label,
		parent: null,
		message,
		detail,
		meta: null,
		entries: [],
	}
}

let _activeContext: LoggerContext<any> | null = null;

export function getLoggerContext(): LoggerContext<any> | null {
	return _activeContext;
}

export function switchLoggerContext(ctx: LoggerContext<any>, scope: LoggerScope<any>): ContextSnapshot | null {
	if (ctx === null) {
		_activeContext = null;
		return null;
	}

	const prev_activeContext = _activeContext;
	const prev_scope = ctx.scope;
	const prev_scopeContext = ctx.scopeContext;
	let prev_scopeContextParent: ScopeEntry | null = null;

	_activeContext = ctx;
	ctx.scope = scope;

	if (ctx.scopeContext) {
		prev_scopeContextParent = ctx.scopeContext.parent;
		ctx.scopeContext.parent = scope.scope();
	}

	return {
		activeContext: prev_activeContext,
		ctx,
		scope: prev_scope,
		scopeContext: prev_scopeContext,
		scopeContextParent: prev_scopeContextParent,
	};
}

export function revertLoggerContext(snapshot: ContextSnapshot) {
	const {
		ctx,
		scopeContext,
	} = snapshot;

	ctx.scope = snapshot.scope;
	ctx.scopeContext = scopeContext;

	if (scopeContext) {
		scopeContext.parent = snapshot.scopeContextParent;
	}

	_activeContext = snapshot.activeContext;
}

export function createLogger<LA extends LoggerAPI>(
	options: Partial<LoggerOptions>,
	factory: (api: LoggerFactoryAPI) => LA,
): Logger<LA> {
	const createCoreLogger = (ctx: LoggerScopeContext) => ({
		add(message: string | Entry, detail?: any): Entry {
			let entry: Entry;

			if (isLogEntry(message)) {
				entry = message;
			} else {
				entry = createEntry(
					'log',
					null,
					null,
					message,
					detail,
				);
			}

			const {detail:args} = entry;

			if (args && args.length && isMeta(args[0])) {
				entry.meta = toMeta(args.shift());
			}

			if (options.time) {
				entry.ts = now();
			}

			const parent = ctx.parent;
			
			if (parent) {
				const length = (entry.parent = parent).entries.push(entry);
				
				if (length > options.storeLast!) {
					parent.entries.splice(0, 1);
				}
			}

			if (options.silent !== true && options.output) {
				options.output(entry);
			}

			return entry;
		},
	});
	const root = createScopeEntry('info', '🚧', '#root', 'root', {
		time,
		created: new Date(),
	});
	const _activeScopeContext: LoggerScopeContext = {
		logger: null,
		parent: root,
	};
	const logger = createCoreLogger(_activeScopeContext);
	const api = factory({
		createEntry,
		logger,
	}) as Logger<LA> & {
		m: EntryMeta | null;
	};

	if (options.silent == null && typeof location !== 'undefined') {
		options.silent = !/^(about:|file:|https?:\/\/localhost\/)/.test(location + '');
	}

	if (options.time == null) {
		options.time = true;
	}

	if (options.storeLast == null) {
		options.storeLast = 1e3;
	}

	_activeScopeContext.logger = logger;

	// Reserved methods
	['add', 'clear', 'scope', 'setup', 'entries', 'last', 'priny', 'm'].forEach((name) => {
		if (api.hasOwnProperty(name)) {
			throw new Error(`[octoLogger] "${name}" is a reserved identifier`);
		}
	});

	api.print = () => {
		function next(root: Entry & {printed?: boolean}) {
			root.printed = false;
			root.entries.forEach((entry: Entry & {printed?: boolean}) => {
				entry.printed = false;
				options.output && options.output(entry);

				if (entry.type === SCOPE) {
					next(entry);
				}
			});
		}

		next(root);

		// Close all groups
		options.output && options.output(null);
	};

	api.add = (...args: any[]) => logger.add(createEntry('log', null, null, null, args));
	api.clear = () => root.entries.splice(0, root.entries.length);
	api.setup = (optionsPatch: Partial<LoggerOptions>) => {
		for (let key in optionsPatch) {
			if (optionsPatch.hasOwnProperty(key)) {
				options[key] = optionsPatch[key];
			}
		}
	};
	api.entries = () => root.entries;
	api.last = () => root.entries[root.entries.length - 1] || null;
	api.scope = function scopeCreator(
		this: Logger<LA>,
		message?: string | Entry,
		detail?: any,
		executor?: ScopeExecutor<LA, LoggerScope<LA>>,
	) {
		const args = arguments;
		let meta: EntryMeta | null = null;

		if (args.length === 0) {
			return (this as any)._scopeEntry;
		}
		
		const firstArg = args[0];
		if (isMeta(firstArg)) {
			meta = toMeta(firstArg);
			message = args[1];
			detail = args[2];
			executor = args[3];
		}

		if (executor == null && typeof detail === 'function') {
			executor = detail;
			detail = null;
		}

		const scopeEntry = isLogEntry(message) ? message : createScopeEntry(
			'info',
			null,
			null,
			message,
			detail,
		);
		
		scopeEntry.meta = meta;
		_activeScopeContext.logger!.add(scopeEntry);

		const logger = createCoreLogger({parent: scopeEntry, logger: null});
		const scopeAPI: any = factory({
			createEntry,
			logger,
		});

		scopeAPI.add = (...args: any[]) => logger.add(createEntry('log', null, null, null, args));
		scopeAPI.scope = scopeCreator;
		(scopeAPI as any)._scopeEntry = scopeEntry;

		// Переключаем scope
		if (typeof executor === 'function') {
			const prev_activeContext = _activeContext;
			const prev_parentEntry = _activeScopeContext.parent;
			const prev_parentLogger = _activeScopeContext.logger;

			_activeContext = {
				entry: scopeEntry,
				scope: scopeAPI as LoggerScope<LA>,
				logger,
				options,
				scopeContext: _activeScopeContext
			};

			_activeScopeContext.parent = scopeEntry;

			executor(scopeAPI);

			_activeContext = prev_activeContext;

			_activeScopeContext.parent = prev_parentEntry;
			_activeScopeContext.logger = prev_parentLogger;
		}

		return scopeAPI;
	};

	return api;
}

const BADGES = {
	log: null,
	info: '❕',
	done: '✅',
	warn: '⚠️',
	error: '🛑',
	verbose: '🔎',
};

export const logger = createLogger({
	output: universalOutput(),
}, ({logger}) => Object.keys(BADGES).reduce((api, level) => {
	api[level] = (...args: any[]) => {
		logger.add(createEntry(level, BADGES[level] || null, null, null, args));
	};

	return api;
}, {}) as {
	[K in (keyof typeof BADGES)]: (...args: any[]) => void;
});
