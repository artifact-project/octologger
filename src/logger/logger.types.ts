import { Output } from '../output/output';

export const ENTRY = 1;
export const SCOPE = 2;

export type CoreLogger = {
	add(entry: Entry): Entry;
	add(message: string, detail?: any): Entry;
}

export type LoggerAPI = {
	[method:string]: (...args: any[]) => any;
}

export type LoggerOptions = {
	meta: boolean;
	time: boolean;
	silent: boolean;
	storeLast: number;
	output: Output | null;
}

export type LoggerFactoryAPI = {
	logger: CoreLogger;
	createEntry: (
		level: Entry['level'],
		badge: Entry['badge'],
		label: Entry['label'],
		message: Entry['message'],
		detail: Entry['detail'],
	) => Entry;
}

export interface Entry {
	cid: number;
	ts: number;
	type: number;
	level: string;
	badge: string | null;
	label: string | null;
	message?: string | null;
	detail: any;
	meta: EntryMeta | null;
	parent: Entry | null;
	entries: Entry[];
}

export type EntryMeta = {
	file: string;
	line: number;
	col: number;
}

export interface ScopeEntry extends Entry {
}

type OctoShell<T>= T & {__octoret__?: never};

export type OctoMethod<T extends (...args: any[]) => any> = T extends (...args: infer A) => infer R
	? (...args: A) => OctoShell<R>
	: OctoShell<T>

export type ScopeExecutor<LA extends LoggerAPI, LS extends LoggerScope<LA>> = (scope: LS) => void;

export type LoggerScope<LA extends LoggerAPI> = {
	[K in keyof LA]: OctoMethod<LA[K]>;
} & {
	add: OctoMethod<(...args: any[]) => Entry>;

	scope(): ScopeEntry; // current scope

	scope(
		message: string | ScopeEntry,
		executor?: ScopeExecutor<LA, LoggerScope<LA>>,
	): OctoShell<LoggerScope<LA>>;

	scope(
		message: string | ScopeEntry,
		detail: any,
		executor?: ScopeExecutor<LA, LoggerScope<LA>>,
	): OctoShell<LoggerScope<LA>>;
}

export type Logger<LA extends LoggerAPI> = LoggerScope<LA> & {
	setup: (optionsPatch: Partial<LoggerOptions>) => void;
	print(): void;
	clear(): Entry[];
	entries(): Entry[];
	last(): Entry;
	// context(): LoggerContext<LA> | null;
}

export type LoggerContext<LA extends LoggerAPI> = {
	scope: LoggerScope<LA>;
	entry: ScopeEntry;
	logger: CoreLogger;
	options: Partial<LoggerOptions>;
	scopeContext: LoggerScopeContext | null;
}

export type LoggerScopeContext = {
	parent: ScopeEntry | null;
	logger: CoreLogger | null;
}

export type ContextSnapshot = {
	activeContext: LoggerContext<any> | null;
	ctx: LoggerContext<any>;
	scope: LoggerScope<any>;
	scopeContext: LoggerScopeContext | null;
	scopeContextParent: ScopeEntry | null;
}

// export type LiteralUnion<
// 	LiteralType,
// 	BaseType extends Primitive
// > = LiteralType | (BaseType & {_?: never});

// export type BaseLogLevels = 
// 	| 'log'
// 	| 'info'
// 	| 'warn'
// 	| 'error'
// 	| 'verbose'
// ;
