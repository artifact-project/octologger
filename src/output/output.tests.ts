import { browserOutput } from './output';
import { createLogger } from '../logger/logger';
import { arrayOutput } from './array';

describe('output', () => {
	it('browser', () => {
		const log: any[] = [];
		const logger = createLogger({
			output: browserOutput({
				out: {
					log: (...args: any[]) => log.push(args),
					group: () => {},
					groupEnd: () => {},
				},
			}),
		}, ({logger, createEntry}) => ({
			test: () => logger.add(createEntry('info', null, 'fail', 'Critical', 123)),
		}));

		logger.test();
		expect(log.length).toEqual(1);
	});

	it('array', () => {
		const log: string[] = [];
		const logger = createLogger({
			time: false,
			output: arrayOutput({out: log}),
		}, () => ({}));

		logger.add('foo', [1, 2], 'and', null, '!==', void 0);
		logger.scope('root', () => {
			logger.add('bar');
			logger.scope('deep', () => {
				logger.add('baz');
				logger.scope('deeper', () => {
					logger.add('wow');
				});
			});
		});
		logger.add('qux');

		expect(log.join('\n')).toMatchSnapshot();
	});
});
