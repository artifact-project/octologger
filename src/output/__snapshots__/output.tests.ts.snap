// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`output array 1`] = `
"foo [1,2] and null !== undefined
root
  | bar
  | deep
     | baz
     | deeper
        | wow
 qux"
`;
